### ISSUE TITLE
GCC backend

### ISSUE THREAD

**Issue by certik** [lfortran/lfortran#1257]
We have LLVM, C, WASM and other backends. One idea is to add a GCC backend.

There are two ways forward.

### C

The easiest way to target GCC is to simply use our C backend.

Pros:

* It has the advantage that we support all GCC versions (past and future)
* We already have it
* No issues with regards to the license of our code
* Easy to develop (no GCC specific learning curve)

Cons:

* It's slightly inefficient (slower compile times) because GCC must parse the C code
* I don't know if there are any possible runtime performance slow downs compared to the native option

### Native

We can add a new backend that takes our ASR and targets one of the GCC internal representations.

Pros:

* It would be faster to compile
* There might be some potential runtime performance benefits 
* The location information might be more accurate
* possibly debugging symbols and experience might be better as well

Cons:

* We would have to write and maintain such a backend and keep it updated and working with all GCC versions
* Probably it would need to be a separate project / fork of GCC and it would need to use the GPL license.

Which GCC internal representation should we target? It seems GFortran has one which we could potentially target; or the [GENERIC](https://gcc.gnu.org/onlinedocs/gccint/GENERIC.html) or [GIMPLE](https://gcc.gnu.org/onlinedocs/gccint/GIMPLE.html#GIMPLE) representations.

**Comment by certik** [comment]
barracuda156 wrote:
> On a side-note, are there any plans for GCC backend?

You can already use GCC via our C backend. In your opinion, what would be the advantage of the native backend?

**Comment by barracuda156** [comment]
@certik `gfortran` seems a logical choice. For versions, I think there is no point in supporting anything earlier than `gcc10` (the last one that can be bootstrapped without C11/C++11 compiler). I think GCC upstream dropped support for gcc versions < 10 too. `gcc10` is likely to be supported in long term for the reason above.
IMO, even starting from the latest `gcc13` is fine: it still builds on old systems and across all archs, just requires an intermediate compiler.

There are some limitations in `gfortran` with regard to FP manipulation and IEEE arithmetic (not implemented for some platforms), but generally speaking GCC should be substantially more consistent and have a better platform coverage. (For example, LLVM is semi-broken on Darwin PPC since its beginning and until present day.)

**Comment by barracuda156** [comment]
> You can already use GCC via our C backend. In your opinion, what would be the advantage of the native backend?

Details here are above my level of expertise, to be honest; I cannot estimate how better performance with GCC-optimized backend might be as compared to a generic C backend.
I had an impression that certain functionality is unavailable unless LLVM is used; if that is the case, it is desirable to implement the same with GCC (since it is more cross-platform). Well, IMO :)

**Comment by certik** [comment]
Thanks for the feedback. Excellent point: "Why do we use LLVM directly if we could just generate C code and use Clang?". The answer to this will then give the answer to whether to use C or native GCC IR.

Here are some of the reasons we use LLVM over Clang:

* LLVM's JIT for the interactive use
* Smaller self-contained LFortran binary (just parts of LLVM that we need, no Clang), everything statically linked
* Faster compilation (no Clang needed to parse C, just LLVM)

GCC also seems to have a [JIT](https://gcc.gnu.org/wiki/JIT) now. Here is an example of Rust using GCC as a backend: https://github.com/rust-lang/rustc_codegen_gcc.

**Comment by barracuda156** [comment]
@certik GCC has JIT for quite a while, AFAICT. Definitely for several major versions. And Rust-GCC is merged into gcc13 (upstream master). I have built it on 10.6.8 for ppc32 few days back, seems to build fine even there.

For Clang, I in no way advocated or implied using it. It is less consistent and more broken on many platforms than even LLVM. It seems to be bothered only with the latest mainstream platforms.

**Comment by certik** [comment]
Good point that Clang has even less platform support than LLVM. It seems the main argument for directly targeting the LLVM or GCC's IR is a more tight integration with their optimizers and backends, not having to worry about possible limitations of their C frontends.
